### PR TITLE
fix: pass CLS transaction to model hooks

### PR DESCRIPTION
**Pull request by heroic**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [ ] Have you added new tests to prevent regressions?
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Fixes: #12973 

### Todos

- [ ] Figure out how and where to write a test case to prevent future regression.

### CODE REVIEW

**Comment by ephys** [comment]
Thank you for the PR :)
Could you add a test in (i think) `test/integration/cls.test.js` that checks the transaction is properly passed where it wasn't before?

**Comment by heroic** [comment]
Hi @ephys I tried! Couldn't figure out how to make it work though. I did test the change manually in our app though. Happy to give it a shot if you can guide.

**Comment by ephys** [comment]
I'll try to take a stab at it next time I can

**Comment by ephys** [comment]
I'm looking into this more and it's weird that your fix is necessary.

`Model#save` calls either `QueryInterface#insert` or `QueryInterface#update` depending on isNewRecord.
Both of these methods then call `Sequelize#query`, which retrieves the transaction from `Sequelize._cls`

**Comment by ephys** [comment]
I've added tests for this in branch `feature/ephys/cls-in-hooks`

Here they are: https://github.com/sequelize/sequelize/blob/feature/ephys/cls-in-hooks/test/integration/cls.test.js#L143
And the commits: https://github.com/sequelize/sequelize/commit/6294414f91692da2432508956389fdf02e33f954

Feel free to copy them in your PR / to cherry pick the commit

It tests every Model hooks:
image

File: test/integration/cls.test.js
Comment (ephys) [inline test/integration/cls.test.js:267]: I made a mistake here
```suggestion
        hooks: ['beforeCreate', 'afterCreate'],
```

**Comment by ephys** [comment]
Well passing the transaction to hooks has been implemented in the #save method but the other model methods aren't implementing it

image

**Comment by heroic** [comment]
> Well passing the transaction to hooks has been implemented in the #save method but the other model methods aren't implementing it
> 
> image

So we'll need to pass the transaction in other functions too. Got it. Will update

**Comment by WikiRik** [comment]
Hi @heroic, did you get a chance to look at this again? Or is there something we can do to help?

**Comment by heroic** [comment]
@WikiRik a colleague of mine is looking to get this closed! We should have this working soon

**Comment by AmirMelio** [comment]
> @WikiRik a colleague of mine is looking to get this closed! We should have this working soon

Hi @heroic , any update?

**Comment by joker00777** [comment]
@AmirMelio I will be continuing it from here.

File: test/integration/cls.test.js
Comment (joker00777) [inline test/integration/cls.test.js:208]: @ephys it should be beforeUpsert i think?

**Review by joker00777 — COMMENTED** [review]
Test file typo

File: test/integration/cls.test.js
Comment (joker00777) [inline test/integration/cls.test.js:288]: these test getting failed the transaction is going in the runhook function but still the test is failing

**Review by joker00777 — COMMENTED** [review]
any idea why these are failing?

File: test/integration/cls.test.js
Comment (ephys) [reply test/integration/cls.test.js]: Yes, that's definitely a typo

File: test/integration/cls.test.js
Comment (ephys) [inline test/integration/cls.test.js:291]:
```suggestion
          this.ParanoidUser = this.sequelize.define('ParanoidUser', {
```

File: test/integration/cls.test.js
Comment (ephys) [inline test/integration/cls.test.js:292]:
```suggestion
            name: DataTypes.STRING,
```

**Review by ephys — CHANGES_REQUESTED** [review]

File: test/integration/cls.test.js
Comment (ephys) [reply test/integration/cls.test.js:288]: I tried running your branch locally but I can't build it.
I think your branch is very outdated, could you look into merging `main` into it?

File: lib/model.js
Comment (joker00777) [reply lib/model.js:4506]: will do.

File: test/integration/cls.test.js
Comment (joker00777) [reply test/integration/cls.test.js:288]: okai

Comment (joker00777) [reply test/integration/cls.test.js:289]: You can try now : )

**Review by joker00777 — COMMENTED** [review]
Sorry for the late response : (

**Comment by WikiRik** [comment]
GitHub has [issues with the caching](https://www.githubstatus.com/incidents/6r4sms3hcj3x) today, I'll rerun the checks after they have solved that incident.

**Review by ephys — CHANGES_REQUESTED** [review]

File: src/model-internals.ts
Comment (joker00777) [reply src/model-internals.ts]: Done

**Comment by joker00777** [comment]
@ephys any updates?

**Comment by joker00777** [comment]
@ephys waiting for your feedback : )

File: test/integration/cls.test.js
Comment (ephys) [inline test/integration/cls.test.js:304]: this test is using `this.ParanoidUser` (and it should), but `testHooks` uses `this.User`

Pass the model to `testHooks` and make `testHooks` use that model instead of `this.User`

```
testHooks({
  getModel: () => this.ParanoidUser,
  // ...
});
```

Note that it has to be a getter like above because `this.ParanoidUser` doesn't exist until `beforeEach` has been executed

File: test/integration/cls.test.js
Comment (ephys) [inline test/integration/cls.test.js:197]: `afterFind` receives the option bug at `optionPos` 1, simply duplicate this test for `afterFind` and set its optionPos at 1

**Review by ephys — CHANGES_REQUESTED** [review]

File: test/integration/cls.test.js
Comment (joker00777) [reply test/integration/cls.test.js:197]: okai

**Comment by joker00777** [comment]
@ephys can we merge it now : ) ?

File: test/integration/cls.test.js
Comment (ephys) [inline test/integration/cls.test.js:147]: this should be an option accepted by `testHooks` instead of a context variable

**Review by ephys — CHANGES_REQUESTED** [review]
There are a few unresolved comments from the previous review :)

File: test/integration/cls.test.js
Comment (ephys) [reply test/integration/cls.test.js]: We're moving away from using `this` in our tests so new code shouldn't rely on it and it'd be better to pass a `getModel` function as an option to the `testHooks` method

**Comment by joker00777** [comment]
Hey @ephys can you take a look now?

**Review by ephys — APPROVED** [review]

**Comment by ephys** [comment]
I think we can try coordinating. If you can click the "update branch" button then @ me, I'll merge this immediately before another PR

**Comment by joker00777** [comment]
@ephys @WikiRik update branch Done : )

**Comment by WikiRik** [comment]
@joker00777 we accidentally merged something else just now, could you update the branch once again?

**Comment by ephys** [comment]
😅

**Comment by joker00777** [comment]
haha done @WikiRik @ephys ; )

**Comment by WikiRik** [comment]
Thanks! After the tests pass, I'll merge this

**Comment by joker00777** [comment]
Thank You : )

**Comment by WikiRik** [comment]
There we go! Thanks again for working on this @joker00777 @heroic

**Pull request merged**
